### PR TITLE
DFCC 1384 ContentToCypher recipes result in different uri in each graph instance

### PR DESCRIPTION
**Pull request by sturatcliffe**

### CODE REVIEW

File: GetJobProfiles/CypherCommandRecipes/CreateOccupationLabelNodes.recipe.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
         {
             "name": "CypherCommand",
             "commands": [
-                "match (o:esco__Occupation)\r\n[token:whereClause]\r\nunwind o.skos__altLabel as altLabels\r\ncreate (al:OccupationLabel:Resource {skos__prefLabel: altLabels, uri: \"<<contentapiprefix>>/occupationlabel/\" + apoc.create.uuid()})\r\ncreate (o)-[:hasAltLabel]->(al)"
+                "with 4 as hexLength\r\nmatch (o:esco__Occupation)\r\n[token:whereClause]\r\nunwind o.skos__altLabel as altLabels\r\nwith o, altLabels, hexLength, '0x' + right( o.uri, hexLength) as baseHexCode\r\nwith o, altLabels, hexLength, baseHexCode, toLower(apoc.text.hexValue( apoc.convert.toInteger(baseHexCode) + apoc.coll.indexOf(o.skos__altLabel,altLabels) + 1)) as replacementChars, left(right( o.uri, 36),36 - hexLength) as trimmedUri, 0 as zero\r\nwith o, altLabels, hexLength, replacementChars, baseHexCode, trimmedUri + case size(replacementChars) when hexLength then replacementChars else 1/zero end as newUri\r\ncreate (al:OccupationLabel:Resource {skos__prefLabel: altLabels, uri: \"<<contentapiprefix>>/occupationlabel/\" + newUri})\r\ncreate (o)-[:hasAltLabel]->(al)"
```
Comment (lazcool) [inline GetJobProfiles/CypherCommandRecipes/CreateOccupationLabelNodes.recipe.json:15]: Looks like the 1/zero is a mechanism to raise an error?
We could instead use `right(replacementChars, hexlength)` to cope with overflow without raising an error.

File: GetJobProfiles/CypherCommandRecipes/CreateSkillLabelNodes.recipe.json [hunk 2]
```diff
@@ -12,9 +12,9 @@
...
-                "match [token:occupationMatch](n:esco__Skill)[token:whereClause]\r\nwith distinct n\r\nunwind n.skos__hiddenLabel as hiddenLabels\r\ncreate (l:SkillLabel:Resource {skos__prefLabel: hiddenLabels, uri: \"<<contentapiprefix>>/SkillLabel/\" + apoc.create.uuid()})\r\ncreate (n)-[:hasHiddenLabel]->(l)"
+                "with 4 as hexLength\r\nmatch [token:occupationMatch](n:esco__Skill)[token:whereClause]\r\nwith distinct n, hexLength\r\nunwind n.skos__prefLabel as prefLabels\r\nwith n, prefLabels, hexLength, '0x' + right( n.uri, hexLength) as baseHexCode\r\nwith n, prefLabels, hexLength, baseHexCode, toLower(apoc.text.hexValue( apoc.convert.toInteger(baseHexCode))) as replacementChars, left(right( n.uri, 36),36 - hexLength) as trimmedUri, 0 as zero\r\nwith n, prefLabels, hexLength, replacementChars, baseHexCode, trimmedUri + case size(replacementChars) when hexLength then replacementChars when hexLength  / 2  then '0000' + replacementChars else 1/zero end as newUri\r\ncreate (l:SkillLabel:Resource {skos__prefLabel: prefLabels, uri: \"<<contentapiprefix>>/SkillLabel/\" + newUri})\r\ncreate (n)-[:hasPrefLabel]->(l)",
```
Comment (lazcool) [inline GetJobProfiles/CypherCommandRecipes/CreateSkillLabelNodes.recipe.json:15]: The case checking for hexLength / 2 looks odd.
Is it to try and handle the replacement chars starting with '0'?
If so, wouldn't it only handle 2 initial 0's, and not say 3 or 1, and shouldn't it be adding 2 0's?

Could it instead pad the start of the value with 0's then take the right of the string, which should also handle overflow?

Should the same logic be used for the occupation label nodes too?

File: GetJobProfiles/CypherCommandRecipes/CreateOccupationLabelNodes.recipe.json [hunk 3]
```diff
@@ -12,7 +12,7 @@
         {
             "name": "CypherCommand",
             "commands": [
-                "match (o:esco__Occupation)\r\n[token:whereClause]\r\nunwind o.skos__altLabel as altLabels\r\ncreate (al:OccupationLabel:Resource {skos__prefLabel: altLabels, uri: \"<<contentapiprefix>>/occupationlabel/\" + apoc.create.uuid()})\r\ncreate (o)-[:hasAltLabel]->(al)"
+                "with 4 as hexLength\r\nmatch (o:esco__Occupation)\r\n[token:whereClause]\r\nunwind o.skos__altLabel as altLabels\r\nwith o, altLabels, hexLength, '0x' + right( o.uri, hexLength) as baseHexCode\r\nwith o, altLabels, hexLength, baseHexCode, toLower(apoc.text.hexValue( apoc.convert.toInteger(baseHexCode) + apoc.coll.indexOf(o.skos__altLabel,altLabels) + 1)) as replacementChars, left(right( o.uri, 36),36 - hexLength) as trimmedUri, 0 as zero\r\nwith o, altLabels, replacementChars, trimmedUri + right (replacementChars,4 ) as newUri\r\ncreate (al:OccupationLabel:Resource {skos__prefLabel: altLabels, uri: \"<<contentapiprefix>>/occupationlabel/\" + newUri})\r\ncreate (o)-[:hasAltLabel]->(al)"
```
[the remark this answers is not in the document]
Comment (SimonWren) [reply GetJobProfiles/CypherCommandRecipes/CreateOccupationLabelNodes.recipe.json:15]: changed the right to use hexLength.
Don't think zero padding is required.  replacementchars will either be 4 chars long or 8 if it overflows

File: GetJobProfiles/CypherCommandRecipes/CreateOccupationPrefLabelNodes.recipe.json [hunk 4]
```diff
@@ -12,7 +12,7 @@
         {
             "name": "CypherCommand",
             "commands": [
-                "match (o:esco__Occupation)\r\n[token:whereClause] unwind o.skos__prefLabel as prefLabel\r\ncreate (al:OccupationLabel:Resource {skos__prefLabel: prefLabel, uri: \"<<contentapiprefix>>/occupationlabel/\" + apoc.create.uuid()})\r\ncreate (o)-[:hasPrefLabel]->(al)"
+                "match (o:esco__Occupation)\r\n[token:whereClause]\r\nwith distinct o\r\nwith o, right( o.uri, 36) as trimmedUri\r\ncreate (l:OccupationLabel:Resource {skos__prefLabel:\r\no.skos__prefLabel, uri:\"<<contentapiprefix>>/occupationlabel/\" + trimmedUri})\r\ncreate (o)-[:hasPrefLabel]->(l)"
```
[the remark this answers is not in the document]
Comment (SimonWren) [reply GetJobProfiles/CypherCommandRecipes/CreateOccupationPrefLabelNodes.recipe.json]: Added the additional logic. Also bumped up the alt labels occupation query to "+2" as index is zero based

File: GetJobProfiles/CypherCommandRecipes/CreateOccupationLabelNodes.recipe.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (lazcool) [reply GetJobProfiles/CypherCommandRecipes/CreateOccupationLabelNodes.recipe.json]: Ah yes apologies, `hexValue` always outputs in sets of 4 chars.

**Review by lazcool — APPROVED** [review]

**Pull request merged**